catalog/rocks/catalog.go

```diff
@@ -0,0 +1,244 @@
+	Committer    string
+	Message      string
+	TreeID       TreeID
+	CreationDate time.Time
+	Parents      []CommitID
+	Metadata     map[string]string
+}
+
+// Branch is a pointer to a commit.
+type Branch struct {
+	CommitID CommitID
+	// nolint: structcheck, unused
+	stagingToken StagingToken
+}
+
+// Diff represents a changed state for a given entry (added, removed, changed, conflict)
+type DiffType uint8
+
+const (
+	DiffTypeAdded DiffType = iota
+	DiffTypeRemoved
+	DiffTypeChanged
+	DiffTypeConflict
+)
+
+type Diff struct {
+	Path Path
+	Type DiffType
+}
+
+// function/methods receiving the following basic types could assume they passed validation
+type (
+	// StorageNamespace is the URI to the storage location
+	StorageNamespace string
+
+	// RepositoryID is an identifier for a repo
+	RepositoryID string
+
+	// Path represents a logical path for an entry
+	Path string
+
+	// Ref could be a commit ID, a branch name, a Tag
+	Ref string
+
+	// TagID represents a named tag pointing at a commit
+	TagID string
+
+	// CommitID is a content addressable hash representing a Commit object
+	CommitID string
+
+	// BranchID is an identifier for a branch
+	BranchID string
+
+	// TreeID represents a snapshot of the tree, referenced by a commit
+	TreeID string
+
+	// StagingToken represents a namespace for writes to apply as uncommitted
+	StagingToken string
+
+	// CommonPrefix represents a path prefixing one or more Entry objects
+	CommonPrefix string
+)
+
+// Listing represents either an entry or a CommonPrefix
+type Listing struct {
+	CommonPrefix
+	*Entry
+}
+
+// Interfaces
+type Catalog interface {
+	// entries
+	GetEntry(ctx context.Context, repositoryID RepositoryID, ref Ref, path Path) (*Entry, error)
+	SetEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path, entry Entry) error
+	DeleteEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path) error
+	ListEntries(ctx context.Context, repositoryID RepositoryID, ref Ref, prefix, from, delimiter string, amount int) ([]Listing, bool, error)
+
+	// refs
+	CreateBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) (Branch, error)
+	GetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (Branch, error)
+	Dereference(ctx context.Context, repositoryID RepositoryID, ref Ref) (CommitID, error)
+	Log(ctx context.Context, repositoryID RepositoryID, commitID CommitID, amount int) ([]Commit, bool, error)
+	ListBranches(ctx context.Context, repositoryID RepositoryID, from BranchID, amount int) ([]Branch, bool, error)
+	DeleteBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+
+	// commits
+	Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, commit Commit) (CommitID, error)
+	Reset(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+	Revert(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) error
```

Comment:
We often confuse Reset/Revert terms, it needs documentation to declare the desired behaviour.